pkg/kubelet/eviction/threshold_notifier_linux.go

```diff
@@ -67,51 +70,118 @@ func NewMemCGThresholdNotifier(path, attribute, threshold, description string, h
 			unix.Close(eventfd)
 		}
 	}()
-	glog.V(3).Infof("eviction: setting notification threshold to %s", threshold)
-	config := fmt.Sprintf("%d %d %s", eventfd, watchfd, threshold)
+	epfd, err := unix.EpollCreate1(0)
+	if err != nil {
+		glog.Warningf("eviction manager: error creating epoll: %v", err)
+		return nil, err
+	}
+	if epfd < 0 {
+		err = fmt.Errorf("EpollCreate1 call failed")
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
```

Comment:
nitpick: I find this approach of checking `if err != nil` in a defer to be error-prone, since it leads to artificial code (`err = fmt.Errorf("EpollCreate1 call failed")`) and can easily break if `err` is not properly managed. Go's scoping can also make this worse.

My suggestion, instead, would be to initialize the fd variables to -1, then install a single defer function like so:

```go
defer func() {
  if watchfd >= 0 {
    unix.Close(watchfd)
  }
  if controlfd >= 0 {
    unix.Close(controlfd)
  }
  if eventfd >= 0 {
    unix.Close(eventfd)
  }
  if epfd >= 0 {
    unix.Close(epfd)
  }
}()
```

And then wrap up the code by setting these all to -1 again after transferring them to the linuxCgroupNotifier:

```go
notifier := &linuxCgroupNotifier{...}
// Prevent deferred cleanup function from closing these fds.
watchfd = -1
controlfd = -1
eventfd = -1
epfd = -1
```

Reply:
I think i'll stick with the current design.  I think it is more readable.

Reply:
Another alternative, that might be more readable than `>=` would be to name the errors 1:1 with the fds.